torch/csrc/distributed/autograd/rpc_messages/rpc_with_profiling_req.cpp

```diff
@@ -0,0 +1,151 @@
+      messageType_(messageType),
+      wrappedRpc_(std::move(wrappedRpc)),
+      wrappedMessageType_(wrappedMessageType),
+      tensors_(std::move(tensors)),
+      profilerConfig_(profilerConfig) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cant be null");
+}
+
+rpc::MessageType RpcWithProfilingReq::wrappedMessageType() const {
+  return wrappedMessageType_;
+}
+
+void RpcWithProfilingReq::setWrappedRpc(
+    std::unique_ptr<RpcCommandBase> wrappedRpc) {
+  wrappedRpc_ = std::move(wrappedRpc);
+}
+
+rpc::Message RpcWithProfilingReq::toMessageImpl() && {
+  // save the original message ID and type before moving it.
+  auto wrappedMsgId = wrappedMessage_.id();
+  auto wrappedMsgType = wrappedMessage_.type();
+  // destructively move the wrappedMessage and get the payload. Now the payload
+  // of wrappedMessage won't be in a valid state.
+  auto wrappedPayload = std::move(wrappedMessage_).movePayload();
+  // The wrapped payload should not be empty
+  TORCH_INTERNAL_ASSERT(
+      !wrappedPayload.empty(), "Wrapped payload should not be empty.");
+  // Create the ivalues to send over. We need to send the original message type
+  // and id, as well as some profiling metadata.
+  // TODO: send profiling key
+  std::vector<at::IValue> ivalues{wrappedMsgType,
+                                  static_cast<int64_t>(profilerConfig_.state),
+                                  profilerConfig_.report_input_shapes,
+                                  profilerConfig_.profile_memory,
+                                  fromWorkerId_};
+  // Pickle it into a char payload to be sent over the wire.
+  std::vector<torch::Tensor> tensorTable;
+  std::vector<char> profilingPayload =
+      jit::pickle(c10::ivalue::Tuple::create(std::move(ivalues)), &tensorTable);
+  // add the profiling payload to the wrapped payload
+  rpc::generateWrappedPayload(wrappedPayload, profilingPayload);
+  // Put the wrapped payload into a message to return.
+  auto returnMsg = rpc::Message(
+      std::move(wrappedPayload),
+      std::move(tensors_),
+      messageType_,
+      wrappedMsgId);
+
+  return returnMsg;
+}
+
+RpcCommandBase& RpcWithProfilingReq::wrappedRpc() {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cannot be null!");
+  return *wrappedRpc_;
+}
+
+torch::autograd::profiler::ProfilerConfig RpcWithProfilingReq::
+    getProfilingConfig() const {
+  return profilerConfig_;
+}
+
+rpc::worker_id_t RpcWithProfilingReq::fromWorkerId() const {
+  return fromWorkerId_;
+}
+
+std::unique_ptr<RpcWithProfilingReq> RpcWithProfilingReq::fromMessage(
+    const rpc::Message& message) {
+  rpc::MessageType origMsgType = message.type();
+  std::vector<torch::Tensor> tensors = message.tensors();
+  int64_t msgId = message.id();
+  auto payload = message.payload();
+  auto tupleElements = rpc::readPayload(payload, message);
+  // Ensure that we have the expected number of elements
+  TORCH_INTERNAL_ASSERT(tupleElements.size() == 5);
```

Comment:
shall we constexpr these const numbers instead of hard-coding them?

Reply:
Sure, although I think it makes sense to do it for the size, not sure if we need to do it for the 0, 1, 2, ... tuple accesses (IMO makes it a bit harder to read)